File: src/lib/snapshot.test.ts
import { describe, it, expect } from "vitest";
import {
  getSnapshotStats,
  formatSnapshotName,
  formatTime,
  compareSnapshots,
} from "./snapshot";
import type { Snapshot } from "../api/snapshot";

describe("snapshot", () => {
  const sampleSnapshot: Snapshot = {
    id: "snap_123",
    path: "C:\\Users\\tester\\.confscope\\backups\\snap_123",
    name: "dev-nacos_public_20240101_100000",
    description: "",
    createdAt: "2024-01-01T10:00:00Z",
    updatedAt: "2024-01-01T10:00:00Z",
    source: {
      connectionId: "conn-1",
      connectionName: "dev-nacos",
      namespace: "public",
      namespaceId: "public",
    },
    configs: [
      {
        dataId: "app.yaml",
        group: "DEFAULT_GROUP",
        content: "server:\n  port: 8080",
        configType: "yaml",
        updateTime: "2024-01-01 10:00:00",
      },
      {
        dataId: "db.properties",
        group: "DEFAULT_GROUP",
        content: "db.url=jdbc:mysql://localhost:3306/test",
        configType: "properties",
        updateTime: "2024-01-02 12:00:00",
      },
      {
        dataId: "app2.yaml",
        group: "GROUP2",
        content: "key: value",
        configType: "yaml",
        updateTime: "2024-01-03 14:00:00",
      },
    ],
  };

  describe("getSnapshotStats", () => {
    it("returns correct stats", () => {
      const stats = getSnapshotStats(sampleSnapshot);
      expect(stats.totalConfigs).toBe(3);
      expect(stats.totalGroups).toBe(2);
      expect(stats.totalNamespaces).toBe(1);
      expect(stats.latestUpdateTime).toBe("2024-01-03 14:00:00");
    });

    it("handles empty configs", () => {
      const emptySnapshot = { ...sampleSnapshot, configs: [] };
      const stats = getSnapshotStats(emptySnapshot);
      expect(stats.totalConfigs).toBe(0);
      expect(stats.latestUpdateTime).toBeNull();
    });
  });

  describe("formatSnapshotName", () => {
    it("returns name when available", () => {
      expect(formatSnapshotName(sampleSnapshot)).toBe("dev-nacos_public_20240101_100000");
    });

    it("returns id when name is empty", () => {
      const noName = { ...sampleSnapshot, name: "" };
      expect(formatSnapshotName(noName)).toBe("snap_123");
    });
  });

  describe("formatTime", () => {
    it("formats date string", () => {
      const result = formatTime("2024-01-15T10:30:00Z");
      expect(result).toContain("2024");
      expect(result).toContain("01");
      expect(result).toContain("15");
    });

    it("formats Date object", () => {
      const result = formatTime(new Date("2024-06-15T14:30:00Z"));
      expect(result).toContain("2024");
      expect(result).toContain("06");
    });
  });

  describe("compareSnapshots", () => {
    it("detects added configs", () => {
      const right = {
        ...sampleSnapshot,
        configs: [
          ...sampleSnapshot.configs,
          {
            dataId: "new.yaml",
            group: "DEFAULT_GROUP",
            content: "new: config",
            configType: "yaml",
            updateTime: "2024-01-04",
          },
        ],
      };
      const diffs = compareSnapshots(sampleSnapshot, right);
      const added = diffs.filter((d) => d.diffType === "added");
      expect(added).toHaveLength(1);
      expect(added[0].dataId).toBe("new.yaml");
    });

    it("detects deleted configs", () => {
      const right = {
        ...sampleSnapshot,
        configs: sampleSnapshot.configs.slice(0, 1),
      };
      const diffs = compareSnapshots(sampleSnapshot, right);
      const deleted = diffs.filter((d) => d.diffType === "deleted");
      expect(deleted).toHaveLength(2);
    });

    it("detects modified configs", () => {
      const right = {
        ...sampleSnapshot,
        configs: sampleSnapshot.configs.map((c) =>
          c.dataId === "app.yaml" ? { ...c, content: "server:\n  port: 9090" } : c
        ),
      };
      const diffs = compareSnapshots(sampleSnapshot, right);
      const modified = diffs.filter((d) => d.diffType === "modified");
      expect(modified).toHaveLength(1);
      expect(modified[0].dataId).toBe("app.yaml");
    });

    it("detects unchanged configs", () => {
      const diffs = compareSnapshots(sampleSnapshot, sampleSnapshot);
      const unchanged = diffs.filter((d) => d.diffType === "unchanged");
      expect(unchanged).toHaveLength(3);
    });
  });
});

File: internal/snapshot/snapshot_test.go
package snapshot

import (
	"os"
	"path/filepath"
	"testing"
)

func TestSnapshotManager_CreateAndList(t *testing.T) {
	tmpDir := t.TempDir()
	mgr := NewSnapshotManager(tmpDir)

	source := SnapshotSource{
		ConnectionID:   "conn-1",
		ConnectionName: "dev-nacos",
		Namespace:      "public",
		NamespaceID:    "public",
	}

	configs := []ConfigSnapshot{
		{
			DataID:     "app.yaml",
			Group:      "DEFAULT_GROUP",
			Content:    "server:\n  port: 8080",
			ConfigType: "yaml",
			UpdateTime: "2024-01-01 10:00:00",
		},
		{
			DataID:     "db.properties",
			Group:      "DEFAULT_GROUP",
			Content:    "db.url=jdbc:mysql://localhost:3306/test",
			ConfigType: "properties",
			UpdateTime: "2024-01-02 12:00:00",
		},
	}

	// 创建快照
	snapshot, err := mgr.CreateSnapshot(source, configs)
	if err != nil {
		t.Fatalf("创建快照失败: %v", err)
	}

	if snapshot.ID == "" {
		t.Error("快照 ID 不能为空")
	}
	if snapshot.Name == "" {
		t.Error("快照名称不能为空")
	}
	if len(snapshot.Configs) != 2 {
		t.Errorf("期望 2 个配置，实际 %d", len(snapshot.Configs))
	}

	// 列出快照
	snapshots, err := mgr.ListSnapshots()
	if err != nil {
		t.Fatalf("列出快照失败: %v", err)
	}
	if len(snapshots) != 1 {
		t.Errorf("期望 1 个快照，实际 %d", len(snapshots))
	}

	// 获取快照
	got, err := mgr.GetSnapshot(snapshot.ID)
	if err != nil {
		t.Fatalf("获取快照失败: %v", err)
	}
	if got.ID != snapshot.ID {
		t.Errorf("快照 ID 不匹配: %s != %s", got.ID, snapshot.ID)
	}
	if len(got.Configs) != 2 {
		t.Errorf("期望 2 个配置，实际 %d", len(got.Configs))
	}
	if got.Configs[0].Content != configs[0].Content {
		t.Errorf("配置内容不匹配")
	}
}

func TestSnapshotManager_Delete(t *testing.T) {
	tmpDir := t.TempDir()
	mgr := NewSnapshotManager(tmpDir)

	source := SnapshotSource{
		ConnectionID:   "conn-1",
		ConnectionName: "dev-nacos",
		Namespace:      "public",
	}

	snapshot, err := mgr.CreateSnapshot(source, nil)
	if err != nil {
		t.Fatalf("创建快照失败: %v", err)
	}

	// 删除快照
	if err := mgr.DeleteSnapshot(snapshot.ID); err != nil {
		t.Fatalf("删除快照失败: %v", err)
	}

	// 验证已删除
	snapshots, err := mgr.ListSnapshots()
	if err != nil {
		t.Fatalf("列出快照失败: %v", err)
	}
	if len(snapshots) != 0 {
		t.Errorf("期望 0 个快照，实际 %d", len(snapshots))
	}
}

func TestSnapshotManager_Validate(t *testing.T) {
	tmpDir := t.TempDir()
	mgr := NewSnapshotManager(tmpDir)

	// 创建有效快照
	source := SnapshotSource{
		ConnectionID:   "conn-1",
		ConnectionName: "dev-nacos",
		Namespace:      "public",
	}

	snapshot, err := mgr.CreateSnapshot(source, []ConfigSnapshot{
		{DataID: "app.yaml", Group: "DEFAULT_GROUP", Content: "test"},
	})
	if err != nil {
		t.Fatalf("创建快照失败: %v", err)
	}

	// 验证有效快照
	snapshotDir := filepath.Join(tmpDir, snapshot.ID)
	if err := mgr.ValidateSnapshot(snapshotDir); err != nil {
		t.Errorf("验证有效快照失败: %v", err)
	}

	// 验证无效目录
	if err := mgr.ValidateSnapshot(filepath.Join(tmpDir, "nonexistent")); err == nil {
		t.Error("期望验证失败，但成功了")
	}

	// 验证缺少 metadata.json 的目录
	invalidDir := filepath.Join(tmpDir, "invalid")
	os.MkdirAll(invalidDir, 0755)
	if err := mgr.ValidateSnapshot(invalidDir); err == nil {
		t.Error("期望验证失败，但成功了")
	}
}
